### ISSUE TITLE
Support for higher order type polymorphism

### ISSUE THREAD

**Issue by jbgi** [functionaljava/functionaljava#126]
Does @functionaljava rules out any support for higher order type polymorphism?
(ie, like how it is done in https://github.com/DanielGronau/highj/wiki)
If yes, what about about a safer, closed-world implementation via GADT, something like:
```java
/**
 * A 'GADT' for "lifting" a type constructors to type parameter level in order to allow the simulation of higher order type
 * polymorphism.
 */
public abstract class h<X, T> {

  private h() {
  }

  private static abstract class Types<R, X, T> {
    R option(final Option<T> option) {
      throw Bottom.undefined();
    }

    R p1(final P1<T> p1) {
      throw Bottom.undefined();
    }

    // ... all generic types we care about
  }

  abstract <R> R match(Types<R, X, T> types);

  public static <T> h<Option<?>, T> option(final Option<T> option) {
    return new h<Option<?>, T>() {
      @Override
      <R> R match(final Types<R, Option<?>, T> types) {
        return types.option(option);
      }
    };
  }

  public static <T> Option<T> narrow(final h<Option<?>, T> hOption) {
    return hOption.match(new Types<Option<T>, Option<?>, T>() {
      @Override
      Option<T> option(final Option<T> option) {
        return option;
      }
    });
  }

}
public abstract class Monad<X> {

  public abstract <A> h<X, A> pure(A a);

  public abstract <A, B> h<X, B> bind(h<X, A> ha, F<A, h<X, B>> f);

}

  public static final Monad<Option<?>> monad = new Monad<Option<?>>() {

    @Override
    public <A> h<Option<?>, A> pure(final A a) {
      return h.option(some(a));
    }

    @Override
    public <A, B> h<Option<?>, B> bind(final h<Option<?>, A> ha, final F<A, h<Option<?>, B>> f) {
      return h.option(h.narrow(ha).bind(a -> h.narrow(f.f(a))));
    }
  };
```

**Comment by gneuvill** [comment]
Hi Jean-Baptiste, all,

> Does @functionaljava rules out any support for higher order type polymorphism?

I hope not ! Indeed, I gave your proposition a shot and here is what I obtained : 

https://gist.github.com/gneuvill/a4a75ef89b04651c0378

which I find (in all humility since this is based on your and @DanielGronau 's idea) rather nice !

Here are the benefits that I perceive from the inclusion of such a feature in FJ : 

- a type class hierarchy expressed in a principled and type safe way
- a new degree of abstraction (@DanielGronau went up to monad transformers ! in java !)
- the amount of code reuse that pertains to the former
- a 100% compatibility with the existing code base (we don't have to break anything)
- and last but not least, a poor man's do notation (which would delight at least me and my colleagues, with regard to the burden of chaining bind, apply and map calls in application code)

Now, granted, there's a glitch : I didn't find a way to avoid casting in the narrowing / coercing functions when dealing with a type constructor of arity 2 or more. That seems to defeat the purpose of the GADT based encoding that you proposed, but not the one of the whole pseudo higher-order types thing, in my opinion. Indeed, the signatures of the narrowing functions ensure type safety nonetheless :  that would thus be a performance glitch only, I think.

So, to recap, a strong :+1:  from me to higher-order type polymorphism in @functionaljava ! (I'm willing to issue a pull request in the weeks to come as a base for discussion)

**Comment by techtangents** [comment]
At first glance this doesn't look type-safe.

**Comment by jbgi** [comment]
@gneuvill the do notation is nice. I take it that, in you gist, ```List<A>``` implements an interface like ```h<List<?>, A>```?
If this is the case then yes, ```narrow``` implementation need to type-cast. This can be alleviated with the GADT encoding above though. But the GADT has the rather big inconvenience to being closed (this is needed to ensure type-safety): not-fj datatypes will not be able to participate in the HKT encoding.
Another way to alleviate some of the risk induced by casting in the ```narrow``` method is to generate it (eg. via Derive4j) but this is still less safe than the GADT encoding.

So there is 3 alternatives:
 1. mourn HKT in java
 2. GADT encoding but closed world
 3. unsafe narrow

Given that Scala and Frege exist on the jvm, with safe HKT, I came to the conclusion that a team mature enough to use HKT would better use one of them than java. I know that, in some organization, using java is mandatory but I think it is best to try to convince management to allow Scala/Frege (or, failing that, to leave the organization asap).
@gneuvill WDYT?

**Comment by gneuvill** [comment]
Hi @techtangents and @jbgi ,

> At first glance this doesn't look type-safe.

Ok, let's take the following example : 

`class Maybe<A> implements __<Maybe<?>, A>`

then indeed, the function

`<A> Maybe<A> narrow(__<Maybe<?>, A> ma)`

is unsafe since one could pass to it anything that implements `__<Maybe<?>, A>` and then boom.

But if we follow @DanielGronau 's highj encoding and tighten it a bit (by making the `µ` witness class protected), we could then have 

```java
@Data(flavour = Flavour.FJ)
public abstract class Maybe<A> implements __<Maybe.µ, A> {
    protected static final class µ {} // protected witness type
    // derive4j stuff
    Maybe() {}
    interface Cases<R, A> {
        R Just(A a);
        R Nothing();
    }
    abstract <R> R match(Cases<R, A> cases);

    public <B> Maybe<B> map(F<A, B> f) {
        return Maybes.<A>cases()
            .Just(a -> Maybes.Just(f.f(a)))
            .Nothing(Maybes::Nothing)
            .f(this);
    }

    public static <A> Maybe<A> narrow(__<Maybe.µ, A> ma) {
        return (Maybe<A>) ma; // no warning from javac
    }

    public static final Instances instances = new Instances();

    public static final class Instances implements Functor<Maybe.µ> {
        private Instances() {}

        @Override
        public <A, B> Maybe<B> fmap(__<Maybe.µ, A> fa, F<A, B> f) {
            return narrow(fa).map(f);
        }
    }
}
```
which I don't immediately see the way you could break it (and the funny thing is that the cast in the narrow function doesn't generate any warning from javac. How is that possible ? Is javac smart enough to deduce that in this case the only possible implementation of `__<Maybe.µ, A>` is `Maybe<A>` ?) 

Your turn now, let me now what you think !

**Comment by jbgi** [comment]
The ```protected``` modifier is superfluous, I think.
 
Still unsafe within ```Maybe``` package:
```java
public class ToTo implements __<Maybe.µ, ToTo> {
  public static void main(String[] args) {
    Maybe.narrow(new ToTo());
  }
}
```
This is still an improvement because we cannot do that if ```ToTo``` is outside ```Maybe``` package so it could be an acceptable trade-off (?).
but then we also lose the ability to reference a ```Maybe``` by it HKT type outside its package, though we probably don't really need to do that (but not sure).

**Comment by gneuvill** [comment]
Jean-Baptiste,

> So there is 3 alternatives:
>  1-  mourn HKT in java

Let's rule it out for now

>  2- GADT encoding but closed world

Actually, I'm not sure I fully grasp your proposition. Could you give an example with the `Either` data type ?
 
> 3- unsafe narrow

Rather 'Unperformant' than unsafe I would say. What do I miss ?

 > Given that Scala and Frege exist on the jvm, with safe HKT, I came to the conclusion that a team mature enough to use HKT would better use one of them than java. I know that, in some organization, using java is mandatory but I think it is best to try to convince management to allow Scala/Frege (or, failing that, to leave the organization asap).

Well that's kind of another discussion, as my request was more focused on the internals of FJ rather than the possible exploitation of the HKT encoding on the client side. I think there are shades of grey between J2EE/Spring/Whatever and haskell on the jvm. And FJ fits that space quite nicely ; in fact, I consider @functionaljava  as an educational tool (because that's through it that I learnt FP), and if it were to include a supplementary artefact to learn  a new concept (HKT), well that would be all for the better, wouldn't it ?

**Comment by jbgi** [comment]
@gneuvill With your improved encoding, possible type-unsafety is limited to package (as noted in my previous comment) so it may be acceptable, but maybe not, given that type-safety is a primary goal of FJ.
As for my GADT encoding, I will need figure out how to generalize it to mutiple type variable (like Either) but, because it involve a lot of wraping/unwrapping I am not a big fan of it my self...

And I agree that HKT would be nice to avoid quite a lot of duplication in FJ code base. 

As a first step, I think the ```__```* classes  (at the root of https://github.com/DanielGronau/highj/tree/master/src/main/java/org/highj) should be package into a separated artefact, so that they could be use separately from highj of FJ. Maybe as sub-project in https://github.com/derive4j for now?

**Comment by gneuvill** [comment]
@jbgi 

> The protected modifier is superfluous

Indeed.

> Still unsafe within Maybe package

Yep. I consider it minor comparing to your other point.

> we also lose the ability to reference a Maybe by it HKT type outside its package

Sh**t. Didn't see that. That's a real problem. That doesn't seem to compromise a type classes hierarchy and the relative do notation though (wherein your only references are in the form of `Monad<f>` or `__<f, A>`)

**Comment by DanielGronau** [comment]
I haven't even tried to get `narrow()` foolproof in highJ. A "hacky" way to improve security would be to require and to ensure that the witness class is inside the implementing class:

```java
public abstract class _<Witness, A> {
    _(Class<Witness> c) {
        if (c.getEnclosingClass() == null ||
           ! c.getEnclosingClass().isAssignableFrom(getClass())) {
            throw new AssertionError();
        }
    }
}
``` 
That means attempts like `ToTo` would immediately explode when created. 

Of course it would be much nicer to have a compile time check, but when I started highJ, I looked quite a while into this, and couldn't find a solution.

**Comment by gneuvill** [comment]
@DanielGronau Thanks for chiming in !

> Of course it would be much nicer to have a compile time check, but when I started highJ, I looked quite a while into this, and couldn't find a solution.

Well, I guess @jbgi would tell you that there is one : an annotation processor ! (such as http://types.cs.washington.edu/checker-framework/ but as matter of fact, @jbgi 's derive4j also performs some compile time checks to properly generate its algebraic data types encoding)

So, as I see it, the new project https://github.com/derive4j/hkt would embed such types as 
```java
@Witnessed // or any other name
public interface __<f, T> {}
```
*PLUS* an annotation processor that would check, for each `@Witnessed` annotated type, that : 
- every extension or implementation of that annotated type declares the following type member : 
```java
public static final class µ {} // possibly declaring a private constructor ?
```
- such an extension or implementation adequately refers to its `µ` member in its extension/implementation declaration  : 
```java
class Toto<A> implements __<Toto.µ, A> {
                              // ^ checked by the apt
  public static final class µ {}
}
```
- possibly generates the corresponding `narrow/coerce` function : 
```java
public static Toto<A> narrow/coerce(__<Toto.µ, A> toto) { return (Toto<A>) toto; }
```

Declaring the annotation processor in a `META-INF/services/javax.annotation.processing.Processor` file inside https://github.com/derive4j/hkt would ensure that it is systematically run by `javac`.

Now, granted, you could still run `javac` with the `apt:none` flag but that would mean that you're strongly determined to destroy your toes with big blasts of sawn-off shotgun, in which case there's not much we can do (as we French use to say "Le client est roi") apart from falling back to a runtime check : 
```java
@Witnessed
public interface __<f, T> {
  static { 
    // do the runtime check in a static block so that it runs at class loading time
  }
}
```
Does all this sound like a credible scenario ?
